src/arm64/arm_detect_encodings.cpp

```diff
@@ -0,0 +1,200 @@
+template<class checker>
+// len is known to be a multiple of 2 when this is called
+int arm_detect_encodings(const char * buf, size_t len) {
+    const char* start = buf;
+    const char* end = buf + len;
+
+    bool is_utf8 = true;
+    bool is_utf16 = true;
+    bool is_utf32 = true;
+
+    int out = 0;
+
+    const auto v_d8 = simd8<uint8_t>::splat(0xd8);
+    const auto v_f8 = simd8<uint8_t>::splat(0xf8);
+
+    uint32x4_t currentmax = vmovq_n_u32(0x0);
+
+    checker check{};
+
+    while(buf + 64 <= end) {
+        uint16x8_t in = vld1q_u16(reinterpret_cast<const uint16_t*>(buf));
+        uint16x8_t secondin = vld1q_u16(reinterpret_cast<const uint16_t*>(buf) + simd16<uint16_t>::SIZE / sizeof(char16_t));
+        uint16x8_t thirdin = vld1q_u16(reinterpret_cast<const uint16_t*>(buf) + 2*simd16<uint16_t>::SIZE / sizeof(char16_t));
+        uint16x8_t fourthin = vld1q_u16(reinterpret_cast<const uint16_t*>(buf) + 3*simd16<uint16_t>::SIZE / sizeof(char16_t));
+
+        const auto u0 = simd16<uint16_t>(in);
+        const auto u1 = simd16<uint16_t>(secondin);
+        const auto u2 = simd16<uint16_t>(thirdin);
+        const auto u3 = simd16<uint16_t>(fourthin);
+
+        const auto v0 = u0.shr<8>();
+        const auto v1 = u1.shr<8>();
+        const auto v2 = u2.shr<8>();
+        const auto v3 = u3.shr<8>();
+
+        const auto in16 = simd16<uint16_t>::pack(v0, v1);
+        const auto nextin16 = simd16<uint16_t>::pack(v2, v3);
+
+        const auto surrogates_wordmask0 = (in16 & v_f8) == v_d8;
+        const auto surrogates_wordmask1 = (nextin16 & v_f8) == v_d8;
+
+        // Check for surrogates
+        if (!surrogates_wordmask0.none() || !surrogates_wordmask1.none()) {
```

Comment:
Under ARM, the 'none()' method is very slow. Please consider switching over to @[danlark1](https://github.com/danlark1)'s  to_bitmask64() which is at least twice as fast... See https://github.com/simdutf/simdutf/pull/145 for my validate_utf16 benchmarks where it made a huge difference.